benches/functional.rs

```diff
@@ -0,0 +1,101 @@
+use bst::functional::Tree;
+use criterion::{
+    black_box, criterion_group, criterion_main, AxisScale, BenchmarkId, Criterion,
+    PlotConfiguration,
+};
+
+/// Returns how many nodes are needed to fill a binary tree with `num_levels` levels.
+fn num_nodes_in_full_tree(num_levels: usize) -> usize {
+    2usize.pow(num_levels as u32) - 1
+}
+
+/// Builds a tree by inserting values in an unbalanced manner. This adds elements in an
+/// ascending manner to ensure the tree is unbalanced (unless it is self-balancing!).
+fn get_unbalanced_tree(num_levels: usize) -> Tree<i32, i32> {
+    let mut tree = Tree::new();
+    let tree_size = num_nodes_in_full_tree(num_levels);
+    for x in (0..).take(tree_size) {
+        tree = tree.insert(x, x);
+    }
+
+    tree
+}
+
+/// Builds a tree by inserting values in a balanced manner. This adds elements so that, without
+/// any self-balancing, the resultant tree will still be balanced.
+///
+/// It ensures there are `num_levels` of nodes, all full.
+fn get_balanced_tree(num_levels: usize) -> Tree<i32, i32> {
+    let tree = Tree::new();
+    let tree_size = num_nodes_in_full_tree(num_levels);
+    let xs = (0..).take(tree_size).collect::<Vec<_>>();
+    fill_balanced_tree(tree, &xs)
+}
+
+/// Recursive helper for [`get_balanced_tree`].
+fn fill_balanced_tree(mut tree: Tree<i32, i32>, xs: &[i32]) -> Tree<i32, i32> {
+    if !xs.is_empty() {
+        let mid = xs.len() / 2;
+        tree = tree.insert(xs[mid], xs[mid]);
+        tree = fill_balanced_tree(tree, &xs[..mid]);
+        tree = fill_balanced_tree(tree, &xs[mid + 1..]);
```

Comment:
For the almost base case where `xs.len() == 1`, `mid` is `0`, and `&xs[mid + 1..]` will be out of bounds, but I assume Rust evaluates that case to an empty slice rather than erroring out?

Reply:
Breadcrumbs for later - `x..` is a `RangeFrom` which implements `std::slice::SliceIndex` - want to see how that gets expanded.